sdk/storage/Azure.Storage.Common/src/StorageServerTimeout.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Azure.Core.Pipeline;
+
+namespace Azure.Storage
+{
+    /// <summary>
+    /// Factory for scopes that apply server timeout to Azure Storage calls.
+    ///
+    /// For more information, see
+    /// <see href="https://docs.microsoft.com/rest/api/storageservices/setting-timeouts-for-blob-service-operations">
+    /// Setting timeouts for Blob service operations</see>,
+    /// <see href="https://docs.microsoft.com/rest/api/storageservices/setting-timeouts-for-file-service-operations">
+    /// Setting timeouts for File service operations</see>,
+    /// <see href="https://docs.microsoft.com/rest/api/storageservices/setting-timeouts-for-queue-service-operations">
+    /// Setting timeouts for Queue service operations</see>.
+    /// </summary>
+    public static class StorageServerTimeout
+    {
+        /// <summary>
+        /// Creates a scope in which all REST requests to Azure Storage would have provided a server timeout.
+        ///
+        /// For more information, see
+        /// <see href="https://docs.microsoft.com/rest/api/storageservices/setting-timeouts-for-blob-service-operations">
+        /// Setting timeouts for Blob service operations</see>,
+        /// <see href="https://docs.microsoft.com/rest/api/storageservices/setting-timeouts-for-file-service-operations">
+        /// Setting timeouts for File service operations</see>,
+        /// <see href="https://docs.microsoft.com/rest/api/storageservices/setting-timeouts-for-queue-service-operations">
+        /// Setting timeouts for Queue service operations</see>.
+        /// </summary>
+        /// <param name="timeoutInSeconds">The server timeout in seconds for each HTTP request.</param>
+        /// <returns>The <see cref="IDisposable"/> instance that needs to be disposed when server timeout shouldn't be used anymore.</returns>
+        /// <example>
+        /// Sample usage:
+        /// <code snippet="Snippet:Sample_StorageServerTimeout">
+        /// BlobServiceClient client = new BlobServiceClient(connectionString, options);
+        /// using (StorageServerTimeout.CreateScope(10))
+        /// {
+        ///     client.GetProperties();
+        /// }
+        /// </code>
+        /// </example>
+        /// <remarks>
+        /// The server timeout is sent to the Azure Storage service for each REST request made within the scope.
+        /// This value is not tracked or validated on the client, it is only passed to the Storage service.
+        /// </remarks>
+        public static IDisposable CreateScope(int? timeoutInSeconds)
+        {
+            return HttpPipeline.CreateHttpMessagePropertiesScope(new Dictionary<string, object> { { Constants.ServerTimeout.HttpMessagePropertyKey, timeoutInSeconds } });
```

Comment:
I feel like we're doing this wrong.  Instead of taking a constant number of seconds, could we compute when we should time out and take the difference in the pipeline so this quantity decreases on each retry?

Reply:
I'm not sure I understand the intent but this sounds more like `MaximumExecutionTime` which now corresponds to providing cancallation token to the api. This [example](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.blob.blobrequestoptions.maximumexecutiontime?view=azure-dotnet-legacy#examples) shows how this was used in previous version.
Could you provide example for the idea you described?

I think the service timeout is a similar primitive know as [NetworkTimeout](https://docs.microsoft.com/en-us/dotnet/api/azure.core.retryoptions.networktimeout?view=azure-dotnet) and applies to each distinct rest call (including retries). So I'd rather see user sizing service timeout similar way they'd size NetworkTimeout (btw. maybe we should also offer way to customize network timeout per operation?).

Overall I'd rather direct customers towards networktimeout, but I can imagine somebody wanting to put a cap on the operation that's applied on the server side to make sure (or rather increase probability) that operation doesn't continue on the other side in background.

For example, I can imagine sizing CTS.CancelAfter to account for all retiries then set NetworkTimeout to max of what I'm expecting for all operation (as this is client scope setting) and then use server timeout if I want some operations to time out before network timeout kicks in (for each retry).

Also, some might not pass cancellation token (max execution time) but still apply network/server timeouts to individual rest calls.

Reply:
I'm for keeping this feature simple. This is a pretty advanced scenario and we should allow customers to have direct access to the feature. If we see there is a need for a more advanced option we can always add it later.